src/cudamatrix/cu-sparse-matrix.h

```diff
@@ -140,79 +122,123 @@ class CuSparseMatrix {
   /// weights; requires indexes.Dim() == weights.Dim().
   /// If trans == kNoTrans, construct a sparse matrix
   /// with num-rows == indexes.Dim() and num-cols = 'dim'.
   /// 'indexes' is expected to contain elements in the
-  /// range [-1, dim - 1].  Each row 'i' of *this after
-  /// calling the constructor will contain zero elements
-  /// if indexes[i] == -1, otherwise it will contain a single
+  /// range [0, dim - 1].  Each row 'i' of *this after
+  /// calling the constructor will contain a single
   /// element at column-index indexes[i] with value weights[i].
   /// If trans == kTrans, the result will be the transpose
   /// of the sparse matrix described above.
-  ///  TODO: implement this.
   CuSparseMatrix(const CuArray<int32> &indexes,
-                 const CuVector<Real> &weights,
-                 int32 dim, MatrixTransposeType trans = kNoTrans);
+                 const CuVectorBase<Real> &weights, int32 dim,
+                 MatrixTransposeType trans = kNoTrans);
 
-  ~CuSparseMatrix() { }
+  ~CuSparseMatrix() {
+    Destroy();
+  }
 
- protected:
+protected:
   // The following two functions should only be called if we did not compile
   // with CUDA or could not get a CUDA card; in that case the contents are
   // interpreted the same as a regular sparse matrix.
   inline const SparseMatrix<Real> &Mat() const {
-    return *(reinterpret_cast<const SparseMatrix<Real>* >(this));
+    return *(reinterpret_cast<const SparseMatrix<Real>*>(this));
   }
   inline SparseMatrix<Real> &Mat() {
-    return *(reinterpret_cast<SparseMatrix<Real>* >(this));
+    return *(reinterpret_cast<SparseMatrix<Real>*>(this));
+  }
+
+  /// Returns pointer to the data array of lenght nnz_ that holds all nonzero
+  /// values in zero-based CSR format
+  const Real* CsrVal() const {
+    return csr_val_;
+  }
+  Real* CsrVal() {
+    return csr_val_;
+  }
+
+  /// Returns pointer to the integer array of length NumRows()+1 that holds
+  /// indices of the first nonzero element in the i-th row, while the last entry
+  /// constans nnz_, as zero-based CSR format is used.
+  const int* CsrRowPtr() const {
+    return csr_row_ptr_col_idx_;
+  }
+  int* CsrRowPtr() {
+    return csr_row_ptr_col_idx_;
+  }
+
+  /// Returns pointer to the integer array of length nnz_ that contains
+  /// the column indices of the corresponding elements in array CsrVal()
+  const int* CsrColIdx() const {
+    return csr_row_ptr_col_idx_ + num_rows_ + 1;
+  }
+  int* CsrColIdx() {
+    return csr_row_ptr_col_idx_ + num_rows_ + 1;
   }
 
- private:
+private:
+  void Destroy();
+
+private:
   // This member is only used if we did not compile for the GPU, or if the GPU
   // is not enabled.  It needs to be first because we reinterpret_cast this
   std::vector<SparseVector<Real> > cpu_rows_;
 
-  MatrixIndexT num_rows_;
-  MatrixIndexT num_cols_;
-
   // This is where the data lives if we are using a GPU.  Notice that the format
```

Comment:
this comment is no longer correct.  I think you should give a pointer to some NVidia or other documentation about how the csr format works, and mention that we store a couple of those arrays in one allocated block of memory.